### PR TITLE
Uncompress request body

### PR DESCRIPTION
**Pull request by ShogunPanda**
This PR adds a mechanism for fastify-compress to handle compressed request payloads.

This is achieved via the new `preDecoding` hook which is still pending in fastify/fastify#2278. If that PR is changed, this PR will be updated accordingly.

Several new options are available, please see README.md changes for more details.

Before merging this PR, we have to update the `package.json` with the right fastift dependency.

Fixes fastify/fastify#1986 and fastify/fastify-compress#105.

Hope this helps!
  Paolo

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -79,6 +124,8 @@ function processParams (opts) {
   }
 
   const supportedEncodings = ['gzip', 'deflate', 'identity']
+
+  /* istanbul ignore next */
```
Comment (mcollina) [inline index.js:128]: I think we can remove the if completely right? we have brotli everywhere?

File: index.js [hunk 2]
```diff
@@ -93,6 +140,51 @@ function processParams (opts) {
...
+      gzip: customZlib.createGunzip || zlib.createGunzip,
+      deflate: customZlib.createInflate || zlib.createInflate
+    },
+    encodings: [],
+    forceEncoding: null
+  }
+
+  const supportedEncodings = ['gzip', 'deflate', 'identity']
+
+  /* istanbul ignore next */
```
Comment (mcollina) [inline index.js:165]: same

File: index.js [hunk 3]
```diff
@@ -174,6 +266,68 @@ function buildRouteCompress (fastify, params, routeOptions, decorateOnly) {
...
+
+    // Check that encoding is supported
+    if (!params.encodings.includes(encoding)) {
+      let errorPayload
+
+      if (params.onUnsupportedRequestEncoding) {
+        try {
+          errorPayload = params.onUnsupportedRequestEncoding(encoding, request)
+        } catch (ex) {
+          /* istanbul ignore next */
```
Comment (mcollina) [inline index.js:300]: Please add a test to cover this case.

File: index.js [hunk 4]
```diff
@@ -240,6 +394,37 @@ function onEnd (err) {
...
+function onDecompressError (request, params, encoding, error) {
+  this.log.debug(`compress: invalid request payload - ${error}`)
+
+  let errorPayload
+
+  if (params.onInvalidRequestPayload) {
+    try {
+      errorPayload = params.onInvalidRequestPayload(encoding, request, error)
+    } catch (ex) {
+      /* istanbul ignore next */
```
Comment (mcollina) [inline index.js:415]: please add a test to cover this.

File: index.js (hunk 3 above)
Comment (ShogunPanda) [reply index.js:300]: Will do

File: index.js (hunk 4 above)
Comment (ShogunPanda) [reply index.js:415]: Will do

File: index.js (hunk 1 above)
Comment (ShogunPanda) [reply index.js:128]: AFAIK no. fastify-compress declares support for Node >= 10.16, while brotli (according to docs) landed in 11.7.0. Am I wrong?

Comment (mcollina) [reply index.js:128]: It got backported in Node 10.16. See https://nodejs.org/docs/latest-v10.x/api/zlib.html

Comment (ShogunPanda) [reply index.js:128]: Oh, nice. I'll remove from both compress and uncompress part then.

File: index.js (hunk 2 above)
Comment (ShogunPanda) [reply index.js]: Done

File: index.js (hunk 4 above)
Comment (ShogunPanda) [reply index.js]: Done

File: index.js (hunk 3 above)
Comment (ShogunPanda) [reply index.js]: Done

File: index.js (hunk 1 above)
Comment (ShogunPanda) [reply index.js]: Done

**Pull request merged**